### PR TITLE
cohere: Add types to clients and drop v4 support from partner package

### PR DESCRIPTION
**Pull request by billytrend-cohere**
cohere: Add types to clients and drop v4 support from partner package

### CODE REVIEW

File: libs/partners/cohere/langchain_cohere/embeddings.py [hunk 1]
```diff
@@ -51,6 +51,18 @@ class Config:
         arbitrary_types_allowed = True
         extra = Extra.forbid
 
+    def get_client(self) -> cohere.Client:
```
Comment (billytrend-cohere) [inline libs/partners/cohere/langchain_cohere/embeddings.py:54]: wondering if we should switch to Field(default_factory=...)

File: libs/partners/cohere/langchain_cohere/embeddings.py [hunk 2]
```diff
@@ -26,9 +26,9 @@ class CohereEmbeddings(BaseModel, Embeddings):
             )
     """
 
-    client: Any  #: :meta private:
+    client: Optional[cohere.Client]  #: :meta private:
```
Comment (efriis) [inline libs/partners/cohere/langchain_cohere/embeddings.py:29]:
```suggestion
    client: cohere.Client = Field(default=None)  #: :meta private:
```

Comment (efriis) [reply libs/partners/cohere/langchain_cohere/embeddings.py:29]: This is how we typically do it. Is there a case where you actually don't want `self.client` to be defined?

Comment (efriis) [reply libs/partners/cohere/langchain_cohere/embeddings.py:29]: Looks like in `validate_environment` it is always initialized anyways, so probably better to just type it as such!

File: libs/partners/cohere/langchain_cohere/llms.py [hunk 3]
```diff
@@ -42,16 +42,18 @@ def acompletion_with_retry(llm: Cohere, **kwargs: Any) -> Any:
...
 
     return _completion_with_retry(**kwargs)
 
 
 class BaseCohere(Serializable):
     """Base class for Cohere models."""
 
-    client: Any = None  #: :meta private:
-    async_client: Any = None  #: :meta private:
+    client: Optional[cohere.Client]  #: :meta private:
```
Comment (efriis) [inline libs/partners/cohere/langchain_cohere/llms.py:53]: see note on embed

File: libs/partners/cohere/langchain_cohere/rerank.py [hunk 4]
```diff
@@ -13,8 +13,8 @@
 class CohereRerank(BaseDocumentCompressor):
     """Document compressor that uses `Cohere Rerank API`."""
 
-    client: Any = None
-    """Cohere client to use for compressing documents."""
+    client: Optional[cohere.Client]  #: :meta private:
```
Comment (efriis) [inline libs/partners/cohere/langchain_cohere/rerank.py:16]: see note on embed

File: libs/partners/cohere/langchain_cohere/embeddings.py (hunk 2 above)
Comment (billytrend-cohere) [reply libs/partners/cohere/langchain_cohere/embeddings.py:29]: that makes sense, will do that pattern, thanks!

**Comment by efriis** [comment]
Do in partner repo!

https://github.com/langchain-ai/langchain-cohere

**Pull request closed without merging**